### ISSUE TITLE
GDScript `range(0, 1, 0.1)` returns an empty array

### ISSUE THREAD

**Issue by tavurth** [godotengine/godot#58210]
### Godot version

v3.4.3.rc1.official [894b6d504]

### System information

Mac OSX (Intel 2019)

### Issue description

Well actually it throws an error if you try and print that, but if you use it in a loop it works fine:

```gdscript
print(range(0.0, 1.0, 0.5))
```
<img width="448" alt="Screenshot 2022-02-17 at 01 32 24" src="https://user-images.githubusercontent.com/100964/154368627-61d6c986-b53c-4f6c-ae95-12d00397c841.png">

```gdscript
for i in range(0.0, 2.0, 0.1):
		print("HERE", i)
```

And `Here` never get's printed.

### Steps to reproduce

Please see above code samples

**Comment by Sauermann** [comment]
According to this comment in the source code, range is supposed to work for integers only:
https://github.com/godotengine/godot/blob/73ece5ba1962eb3c19ac529b8cb4ac18aefd5cac/modules/gdscript/gdscript_analyzer.cpp#L1243-L1244

Here is the location, where they are converted:
https://github.com/godotengine/godot/blob/73ece5ba1962eb3c19ac529b8cb4ac18aefd5cac/modules/gdscript/gdscript_analyzer.cpp#L1289-L1293

The documentation doesn't say anything about `range` being used for integers only - at least I didn't find anything about it.
v4.0 also treats `range` as integers.

**Comment by novaplusplus** [comment]
I suppose an "frange" function could be useful, although of course the third parameter would be mandatory to know the increment.

**Comment by akien-mga** [comment]
Precision errors accumulate on floats, so it's much better to keep integers for your iterator. Just divide it by whatever you need to have the float range you want, e.g.:
```gdscript
for i in range(20):
	print("HERE", i / 20.0)
```

That being said, using a float step in `range()` should probably raise a warning instead of just coercing the values silently.

**Comment by tavurth** [comment]
> Precision errors accumulate on floats

Unless we were to floor it each step as in the example above, it would be a bit easier for junior devs

Perhaps it would be possible to do this for a `frange` util instead though to save performance for range

**Comment by novaplusplus** [comment]
Yeah. "frange" would be integers internally, only returning floats. It just does a bit of math behind the scenes to set up the array.

**Comment by tavurth** [comment]
I think the problem is that it just doesn't run without any warning that it's flooring to int.

This caused me a hard to track down bug with the range in the title because it was silently failing for some values.

**Comment by name-here** [comment]
> > That being said, using a float step in range() should probably raise a warning instead of just coercing the values silently.
> 
> Godot does not show a warning while coercing from float to int.

Honestly, the better fix might be to always give a warning for coercing from float to int.  Many languages don't allow implicit conversion from float to int at all, since it can cause accidental truncation like we see here.

**Comment by Calinou** [comment]
> Honestly, the better fix might be to always give a warning for coercing from float to int. Many languages don't allow implicit conversion from float to int at all, since it can cause accidental truncation like we see here.

There is already such a warning, but it's not always triggered: https://github.com/godotengine/godot/blob/6a51999b7f4d9f7b1f950d7a655a464f28542318/modules/gdscript/gdscript_warning.cpp#L88

**Comment by vnen** [comment]
We can add a specific warning for this case, or maybe even an error since I don't think there's a valid case for using floats in `range()`

**Comment by cdemirer** [comment]
Since
```gdscript
for i in Vector3(0.0, 5.0, 1.5):
	print(i)
```
```
0
1.5
3
4.5
```
works, wouldn't it make sense to have a `range` equivalent?

**Comment by tavurth** [comment]
> ```gdscript
> for i in Vector3(0.0, 5.0, 1.5):
> 	print(i)
> ```

Wow I didn't know about this. 
Seems aliasing `range` to `vector3` can be a solution if it covers all cases 😅

**Comment by novaplusplus** [comment]
> Having this feature built into the language but only to those who know about it also doesn't seem so kind to the developer. If I start to use `Vector3` in my loops then others who come along later probably won't understand at once.

I know my first response was "wait, what the hell?"

That's such an odd and obscure syntax. If it weren't for how specifically it works I would almost be inclined to say it's an accident.
